subsys/net/conn_mgr/conn_mgr_monitor.c

```diff
@@ -141,10 +152,17 @@ static void conn_mgr_mon_handle_update(void)
 		if (ready_count == 0) {
 			/* We just lost connectivity */
 			net_mgmt_event_notify(NET_EVENT_L4_DISCONNECTED, last_iface_down);
+
+			conn_mgr_refresh_online_connectivity_check();
 		} else if (original_ready_count == 0) {
 			/* We just gained connectivity */
 			net_mgmt_event_notify(NET_EVENT_L4_CONNECTED, last_iface_up);
 		}
+
+		if (ready_count > 0) {
+			/* Trigger online check to be done for each ready interface */
+			conn_mgr_trigger_online_connectivity_check();
+		}
```

Comment:
Currently the `conn_mgr_refresh_online_connectivity_check();` generates offline event and then re-triggers online check. There are some extra checks in `conn_mgr_online_connectivity_check()` for not triggering online check unnecessarily.
Also the `conn_mgr_trigger_online_connectivity_check()` does initialization if not done already as this might be the first time the online check functions are called.

Reply:
My only point here is instead of having two separate places where you conditionally call `conn_mgr_trigger_online_connectivity_check`, you could just ***always*** call `conn_mgr_trigger_online_connectivity_check` at the end, since `conn_mgr_trigger_online_connectivity_check` is always called anyways


The way things are written now, it gets called if `ready_count==0`, and it also gets called if `ready_count>0`, and since ready_count cannot be negative, that means it just always gets called

Reply:
Ah, you are right, I've been misreading `conn_mgr_refresh_online_connectivity_check` and `conn_mgr_trigger_online_connectivity_check` as the same function. I think the function names should at a minimum be updated to be a bit closer to their actual function.


I am concerned, because it seems to me, then, that the only way `NET_EVENT_CONNECTIVITY_OFFLINE` can get fired is if all ifaces are unready.


Wouldn't you want `NET_EVENT_CONNECTIVITY_OFFLINE` to fire in the edge case I pointed out?

Reply:
> that the only way `NET_EVENT_CONNECTIVITY_OFFLINE` can get fired is if all ifaces are unready.

Hmm, would that be what we want here? The offline event is generated if there is no online connectivity any more.
I have envisioned this so that if there is one online interface, there is no need to trigger another online check at all. As we have only one host / URL in the configuration to check, it should not matter what interface we check.

Reply:
> The offline event is generated if there is no online connectivity any more.

I am trying to point out that it is possible for there to be ready ifaces, but for there to be no online connectivity.

> As we have only one host / URL in the configuration to check, it should not matter what interface we check.

As long as one iface can reach that host, yes -- but it is possible for there to be at least one oper-up (ready) iface while none of the ifaces are passing online check.

Reply:
I'll reiterate the edge-case scenario -- Imagine we are dealing with a device with two Wi-Fi interfaces, WIFIA an WIFIB.

Imagine there are two networks, NETWORKA, and NETWORKB. NETWORKA is a public network that can reach  the internet. NETWORKB is a private internal network that cannot reach the internet.

On boot, WIFIA connects to NETWORKA and becomes Oper-Up (ready). WIFIB also connects to NETWORKB and becomes oper-up (ready).

This triggers online checks. The online check for WIFIB will fail, since it cannot contact the internet. But the online check passes for WIFIA, since it can.

Now, later suppose someone accidentally unplugs the AP for NETWORKA. The WIFIA iface will detect it is no longer on that network and go oper-down.

In this scenario, the `NET_EVENT_CONNECTIVITY_OFFLINE` ***should*** be fired, since we no longer can contact the internet.

But it ***won't*** ever be fired unless WIFIB also goes down

Reply:
I think we handle this case just fine with current code:

Lets assume A and B are in ready, and A is online. If now A goes down (step 1. in code snippet below), we should be able to catch that by `conn_mgr_online_connectivity_check()` which checks the states and if the interface was in READY but is no longer, then we check the online status for that interface in step 2)
```
	for (i = 0; i < state_count; i++) {
1)->		if (states[i] & CONN_MGR_IF_READY) {
			k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);

			if (!(states[i] & CONN_MGR_IF_READY) ||
			    (states[i] & CONN_MGR_IF_ONLINE_CHECK)) {
				k_mutex_unlock(&conn_mgr_mon_lock);
				continue;
			}

			k_mutex_unlock(&conn_mgr_mon_lock);

			/* Do connectivity check and if it succeeds,
			 * generate CONNECTIVITY_ONLINE event. The online
			 * check might take some time.
			 */
			k_mutex_lock(&check_running_lock, K_FOREVER);

			if (check_running) {
				NET_DBG("Online check already running");
				k_mutex_unlock(&check_running_lock);
				break;
			}

			check_running = true;

			k_mutex_unlock(&check_running_lock);

			k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);

			/* Tells that we are doing online check for this interface */
			states[i] |= CONN_MGR_IF_ONLINE_CHECK;

			k_mutex_unlock(&conn_mgr_mon_lock);

			/* Did we get to online? This is a synchronous call and
			 * in worst case might take a long time to return.
			 */
			ret = do_online_check(conn_mgr_mon_get_if_by_index(i));
			if (ret == 0) {
				/* We are online, no need to check the other interfaces */
				break;
			}

			k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);

			states[i] &= ~CONN_MGR_IF_ONLINE_CHECK;

			k_mutex_unlock(&conn_mgr_mon_lock);
		} else {
			bool trigger_check = false;

			k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);

			/* If this interface was online, then we should re-check */
2)->			if (states[i] & CONN_MGR_IF_ONLINE_CHECK) {
				trigger_check = true;
			}

			/* Clear the online check flag for non ready interface */
			states[i] &= ~CONN_MGR_IF_ONLINE_CHECK;

			k_mutex_unlock(&conn_mgr_mon_lock);

			if (trigger_check) {
				conn_mgr_refresh_online_connectivity_check();
				break;
			}
		}
	}
```
Or did I miss something?

Reply:
Ah, yes, `conn_mgr_refresh_online_connectivity_check` does fire `NET_EVENT_CONNECTIVITY_OFFLINE`. I keep forgetting that. I think a different name for `conn_mgr_refresh_online_connectivity_check` might really make sense.

But now I am a little bit confused. Won't `conn_mgr_refresh_online_connectivity_check` get fired every time an online iface goes offline?

What if there are two online ifaces, and one goes down? In that case you would NOT want fire `NET_EVENT_CONNECTIVITY_OFFLINE`, right?

Reply:
There cannot be two online interfaces, that is the whole point. If we have found out that one of the interfaces is online, no point doing another check.

Reply:
Ahhhhh, alright. I apologize, I severely misinterpreted the intent, here. You are deliberately only ever marking a single iface as "Online" at a time.

In that case, I think this logic all works as you intend, you're right.

This makes me realize, though, that I was wrong to suggest that you use the conn_mgr states array for tracking this.

I recommend, instead, that you keep a global pointer to the iface who should be marked as `CONN_MGR_IF_ONLINE_CHECK`

This would have several advantages:

1. No longer any need to access the conn_mgr states[] array outside core code (I am pretty sure -- if not, I am sure the conn_mgr public API could be adjusted slightly to allow this)
2. Doing this would possibly let you make connectivity checks a totally separate module, too -- instead of directly firing the connectivity check functions in the conn_mgr event loop, you could instead subscribe to the L4_DISCONNECTED event to trigger `conn_mgr_trigger_online_connectivity_check`. Then you could. I'm pretty sure, just use iface oper_up and oper_down events to trigger `conn_mgr_trigger_online_connectivity_check`. But if you really want to wait until conn_mgr determines the total number of enabled ifaces has changed, we could honestly just expose that as a new event.

I do think it makes sense to keep this feature inside the conn_mgr module, but having it ***function*** as a separate component via event listeners would be a lot cleaner and a lot more modular IMO.

Reply:
This PR has been pending for a long time, let me try to resurrect it.

> This makes me realize, though, that I was wrong to suggest that you use the conn_mgr states array for tracking this.

If you are referring to `states` array and the `CONN_MGR_IF_ONLINE_CHECK` bit, I think we need it in order to know which interfaces were done online check. Using a global pointer does not really cut it.

Anyway, I might have missed something as this has been too long time in stale state. I will re-submit this against latest main and lets try to get this re-reviewed and finally merged.

Reply:
Sorry I missed your reply, here!

Alright, I will re-review from scratch when you request my review